kcl/src/main/scala/kinesis4cats/kcl/KCLConsumer.scala

```diff
@@ -175,6 +178,84 @@ object KCLConsumer {
     )(cb)
     .map(new KCLConsumer[F](_))
 
+  def builder[F[_]: Async](
+      appName: String,
+      streamTracker: StreamTracker
+  ): KCLConsumerBuilder[F] =
+    new KCLConsumerBuilder[F](
+      appName = appName,
+      streamTracker = streamTracker,
+      mkWorkerId = UUIDGen[F].randomUUID.map(_.toString),
+      kinesisClientResource = Resource
+        .fromAutoCloseable(
+          Sync[F].delay(KinesisAsyncClient.builder().build())
+        ),
+      dynamoClientResource = Resource.fromAutoCloseable(
+        Sync[F].delay(DynamoDbAsyncClient.builder().build())
+      ),
+      cloudWatchClientResource = Resource.fromAutoCloseable(
+        Sync[F].delay(CloudWatchAsyncClient.builder().build())
+      ),
+      callback = Function.const(Applicative[F].unit),
+      config = identity
+    )
+
+  final case class KCLConsumerBuilder[F[_]: Async] private[KCLConsumer] (
+      private val appName: String,
+      private val streamTracker: StreamTracker,
+      private val mkWorkerId: F[String],
+      private val kinesisClientResource: Resource[F, KinesisAsyncClient],
+      private val dynamoClientResource: Resource[F, DynamoDbAsyncClient],
+      private val cloudWatchClientResource: Resource[F, CloudWatchAsyncClient],
+      private val callback: List[CommittableRecord[F]] => F[Unit],
+      private val config: Config[F] => Config[F]
+  ) {
+    def withKinesisClient(
+        kinesisClient: KinesisAsyncClient
+    ): KCLConsumerBuilder[F] =
+      copy(kinesisClientResource = Resource.pure(kinesisClient))
+
+    def withDynamoClient(
+        dynamoClient: DynamoDbAsyncClient
+    ): KCLConsumerBuilder[F] =
+      copy(dynamoClientResource = Resource.pure(dynamoClient))
+
+    def withCloudWatchClient(
+        cloudWatchClient: CloudWatchAsyncClient
+    ): KCLConsumerBuilder[F] =
+      copy(cloudWatchClientResource = Resource.pure(cloudWatchClient))
+
+    def withWorkerId(workerId: String): KCLConsumerBuilder[F] =
+      copy(mkWorkerId = Applicative[F].pure(workerId))
+
+    def withCallback(
+        cb: List[CommittableRecord[F]] => F[Unit]
+    ): KCLConsumerBuilder[F] =
+      copy(callback = cb)
+
+    def configure(f: Config[F] => Config[F]): KCLConsumerBuilder[F] =
+      copy(config = f)
+
+    def build(implicit
+        encoders: RecordProcessor.LogEncoders
+    ): Resource[F, KCLConsumer[F]] = for {
```

Comment:
**thought:** given LogEncoders is not a typeclass, I think it might be another candidate for something that could be passed explicitly, with a default (Show) that can be overridden using the builder.

Reply:
I would be fine with that, that would be one less import. I like that. It would be a large change but I'd accept it.